website/docs/guides/migration/from-v2-to-v3.md

```diff
@@ -0,0 +1,305 @@
+
+Before, the number of workers was determined statically by configuration at the time of application startup. With KafkaFlow v3, a new method was introduced that supports calculating the number of available workers periodically. This allows the application to adapt constantly and also being able to scale the number of workers during the application's lifetime.
+
+To accommodate this requirement multiple changes were made to the core of KafkaFlow, next we will go through some of these changes.
+
+### 1.1 Dynamic Worker Pool Scaling Configuration
+
+> Related Issues: [#429](https://github.com/Farfetch/kafkaflow/issues/429)
+
+Two new overrides were added to the consumer configuration. Nevertheless, the previous method of statically determining the worker count is still available.
+
+The new overrides added were:
+
+- `WithWorkersCount(Func<WorkersCountContext, IDependencyResolver, Task<int>> calculator)`
+- `WithWorkersCount(Func<WorkersCountContext, IDependencyResolver, Task<int>> calculator, TimeSpan evaluationInterval)`
+
+The difference between both overrides is that the first works with a predefined frequency of 5 minutes between calls to the delegate that returns the number of workers. While in the second override that frequency can be user-defined.
+
+Additionally, the delegate receives a `WorkersCountContext` that sends some consumer-related context to the callback like the Consumer Name, the Consumer Group Key and the Consumer Partitions Assignment. Also, the `IDependencyResolver` is made available in case some dependencies need to be resolved to determine the worker count.
+
+This means that a consumer worker count can be configured like this:
+
+```csharp
+.AddConsumer(
+    consumerBuilder => consumerBuilder
+        .Topic("test-topic")
+        .WithGroupId("group1")
+        .WithWorkersCount((context, resolver) => {
+          // Use whatever logic to determine the worker count
+          return GetWorkerCount(context, resolver);
+        }, TimeSpan.FromMinutes(15))
+)
+```
+
+### 1.2 Improved Mechanism for Signalling Message Processing Completion
+
+> Related Issues: [#426](https://github.com/Farfetch/kafkaflow/issues/426)
+
+For the KafkaFlow engine to adapt to the dynamic change of workers, the way the messages are signaled as completed also needed to be updated because the workers can only be scaled after all the messages in the worker's buffers are signaled as completed.
+
+To achieve this, a new method in the message context was provided to signal the message completion `.Complete()`. Additionally, a new property called `AutoMessageCompletion` was added to signal if KafkaFlow should signal the message as completed automatically which is the default behavior. This property is useful for scenarios where we don't want to complete the message right away, for example in Batch Consume scenarios.
+
+### 1.3 Expose Worker Events to Client Applications
```

Comment:
This was already included in version 2.5
https://github.com/Farfetch/kafkaflow/pull/451

Reply:
However, you can say that the tracing consumer events are now working with batch consumers, which was not the case in version 2.
I removed that line in the documentation regarding the GlobalEvents in the PR that adapted them to version 3
https://github.com/Farfetch/kafkaflow/pull/460/files#diff-6901dcd769d2a84ca151a31802de6971e5b6acd9bfded7b8085bc2631e08047c

Reply:
We discussed that [here](https://github.com/Farfetch/kafkaflow/pull/471#discussion_r1400170791).

I believe that being something that was just recently introduced, it can be a way to communicate to kafkaflow users this feature which can be very useful to them.